jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,353 @@
+     */
+    int size();
+
+    /**
+     * Get a {@link Set} of the keys contained in this index.
+     *
+     * @return a {@link Set} of the keys contained in this index.
+     */
+    Set<String> keySet();
+
+    /**
+     * A mutable String lookup data structure.
+     * Implementations are not thread-safe.
+     * @param <V> the entry type
+     */
+    interface Mutable<V> extends Index<V>
+    {
+        /**
+         * Put an entry into the index.
+         *
+         * @param s The key for the entry
+         * @param v The value of the entry
+         * @return True if the index had capacity to add the field.
+         */
+        boolean put(String s, V v);
+
+        /**
+         * Put a value as both a key and a value.
+         *
+         * @param v The value and key
+         * @return True if the Trie had capacity to add the field.
+         */
+        boolean put(V v);
+
+        /**
+         * Remove an entry from the index.
+         *
+         * @param s The key for the entry
+         * @return The removed value of the entry
+         */
+        V remove(String s);
+
+        /**
+         * Remove all entries from the index.
+         */
+        void clear();
+
+        /**
+         * Builder of {@link Index.Mutable} instances. Such builder cannot be
+         * directly created, it is instead returned by calling {@link Index.Builder#mutable()}.
+         * @param <V> the entry type
+         */
+        class Builder<V> extends Index.Builder<V>
+        {
+            private int maxCapacity = -1;
+
+            Builder(boolean caseSensitive, Map<String, V> contents)
+            {
+                super(caseSensitive, contents);
+            }
+
+            /**
+             * Configure a maximum capacity for the mutable index.
+             * A negative value means there is no capacity limit and
+             * the index can grow without limits.
+             * The default value is -1.
+             * @param capacity the maximum capacity of the index.
+             * @return this
+             */
+            public Builder<V> maxCapacity(int capacity)
+            {
+                this.maxCapacity = capacity;
+                return this;
+            }
+
+            /**
+             * Build a {@link Mutable} instance.
+             * @return a {@link Mutable} instance.
+             */
+            public Mutable<V> build()
+            {
+                if (contents != null && maxCapacity == 0)
+                    throw new IllegalStateException("Cannot create a mutable index with maxCapacity=0 and some contents");
+
+                Mutable<V> result;
+                if (maxCapacity > 0)
+                    result = new ArrayTernaryTrie<>(!caseSensitive, maxCapacity);
+                else if (maxCapacity < 0)
+                    result = new ArrayTernaryTrie.Growing<>(!caseSensitive, 512, 512);
+                else
+                    result = EmptyTrie.instance(caseSensitive);
```

Comment:
Is there a reason to use `ArrayTrie` in any cases?

Reply:
I think if the try is immutable with known content, then we can use ArrayTrie rather than ArrayTernaryTrie.... althought there is also the case of knowing the alphabet used.... ie if non US ASCII characters are used.... so perhaps we need to scan for them....   Hmmm after this PR I think

Reply:
As discussed, we need to improve the algorithm that chooses the underlying implementation.